### PR TITLE
feat(Architecture): add methods to get difference of two architecture object

### PR DESCRIPTION
**Pull request by taro-yu**
Signed-off-by: yu-taro- <milktea1621@gmail.com>

## Description

Implemented diff_node_names

## Related links
https://tier4.atlassian.net/browse/T4PB-18417

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] (Optional) The PR has been properly tested with CARET_report verification.
- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

**Comment by taro-yu** [comment]
- Add class Diff.
- Function diff_nodes_names implemented using set.
- Add test with mocker.

**Comment by rokamu623** [comment]
@taro-yu 
For basic matters,
- Please follow the rules of commit message (e.g., feat:, fix:, etc.)
- Please --signoff when you do commit.
- Please merge main into this commit.

File: src/test/architecture/test_architecture.py [hunk 1]
```diff
@@ -312,3 +312,11 @@ def test_verify_callback_uniqueness(self, mocker, caplog):
...
         Architecture._verify([node_mock])
         assert len(caplog.record_tuples) == 1
+
+    def test_diff_node_names(self, mocker):
+        node_mock = mocker.Mock(spec=Architecture)
+        node_mock2 = mocker.Mock(spec=Architecture)
+        mocker.patch.object(node_mock, 'node_names', 'hoge')
+        mocker.patch.object(node_mock2, 'node_names', 'hoge')
+        name1, name2 = Architecture._diff_node_names(node_mock, node_mock2)
+        assert name1 == name2
```
Comment (rokamu623) [inline src/test/architecture/test_architecture.py:322]: The type of the variable patched in `node_names` is incorrect.
The type of `node_names` is `Tuple[str]`, but the type of `hoge` is `str`.

Comment (rokamu623) [inline src/test/architecture/test_architecture.py:322]: Please add the following test pattern.
- node name only in node_mock
- node name only in node_mock2

Comment (rokamu623) [inline src/test/architecture/test_architecture.py:322]: On the right side of `assert`, write the expected correct value.

in this case,
```
assert name1 == ('hoge', )
```

File: src/test/architecture/test_architecture.py [hunk 2]
```diff
@@ -312,3 +312,11 @@ def test_verify_callback_uniqueness(self, mocker, caplog):
         caplog.clear()
         Architecture._verify([node_mock])
         assert len(caplog.record_tuples) == 1
+
+    def test_diff_node_names(self, mocker):
+        node_mock = mocker.Mock(spec=Architecture)
+        node_mock2 = mocker.Mock(spec=Architecture)
```
Comment (rokamu623) [inline src/test/architecture/test_architecture.py:318]: This is not mock of NodeStruct.
Please change the naming.

ex)
- architecture_mock0

**Review by rokamu623 — CHANGES_REQUESTED** [review]
Here are some comments on the entire implementation.

- Do not use `@staticmethod`.
  - For easier tests
```
def diff_node_names(self, arch1, arch2):
  diff = Diff(arch1, arch2)
  return diff.diff_node_names
```
- Create a new py file for `Diff class`.
- Please also address the minor problems commented below.

**Review by rokamu623 — CHANGES_REQUESTED** [review]
The problem as a whole is described here.

- [ ] Please do not use `@staticmethod`.
```
def diff_by_nodes(arch1, arch2):
  return Diff(arch1, arch2).diff_by_nodes
```
- [ ] Please create a class for the Diff test.

File: src/caret_analyze/architecture/architecture.py [hunk 3]
```diff
@@ -297,3 +295,23 @@ def update_named_path(self, path_name: str, path_info: PathStructValue):
...
+        pass
+
+    @staticmethod
+    def _diff_node_names(
+        arch1: Architecture,
+        arch2: Architecture
+    ) -> Tuple[Tuple[str, ...], Tuple[str, ...]]:
+        set_arch1 = set(arch1.node_names)
+        set_arch2 = set(arch2.node_names)
+        inter_arch1_arch2 = set_arch1 & set_arch2
```
Comment (rokamu623) [inline src/caret_analyze/architecture/architecture.py:314]: - [ ] It appears to be a redundant description.

```suggestion
        inter_arch1_arch2 = set(arch1.node_names) & set(arch2.node_names)
```

**Review by rokamu623 — CHANGES_REQUESTED** [review]

File: src/caret_analyze/architecture/architecture.py (hunk 3 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/architecture.py]: After discussion, it is determined that the current implementation is acceptable.

**Comment by taro-yu** [comment]
@rokamu623 
add diff_node_topics/pubs/subs/callbacks() and make mock tests for these  funcs.

File: src/caret_analyze/architecture/architecture.py [hunk 4]
```diff
@@ -466,6 +466,41 @@ def rename_topic(self, src: str, dst: str) -> None:
...
 
+    @staticmethod
+    def diff_node_names(
+            left_arch: Architecture,
+            right_arch: Architecture
+    ) -> Tuple[Tuple[str, ...], Tuple[str, ...]]:
+        return DiffArchitecture(left_arch, right_arch).diff_node_names()
+
+    @staticmethod
+    def diff_node_topics(
```
Comment (rokamu623) [inline src/caret_analyze/architecture/architecture.py:477]: This function compares all topic in Architecture.
Therefore, the name of this function should be,
```suggestion
    def diff_topic_names(
```

DiffArchitecture.diff_node_topics, too.

File: src/test/architecture/test_architecture.py [hunk 5]
```diff
@@ -1019,3 +1019,140 @@ def test_rename_path(self, mocker):
...
+
+
+class TestDiff:
+
+    def test_diff_node_names(self, mocker):
+        architecture_mock1 = mocker.Mock(spec=Architecture)
+        architecture_mock2 = mocker.Mock(spec=Architecture)
+        mocker.patch.object(architecture_mock1,
+                            'node_names',
+                            ('drive_node', 'actuator_dummy_node'))
```
Comment (rokamu623) [inline src/test/architecture/test_architecture.py:1031]: Use an abstract name.
ex) `node1`

File: src/test/architecture/test_architecture.py [hunk 6]
```diff
@@ -1019,3 +1019,140 @@ def test_rename_path(self, mocker):
...
+        mocker.patch.object(architecture_mock2,
+                            'node_names',
+                            ('drive_node', 'filter_node', 'message_driven_node'))
+        name1, name2 = Architecture.diff_node_names(architecture_mock1, architecture_mock2)
```
Comment (rokamu623) [inline src/test/architecture/test_architecture.py:1035]: For the entire test code:

The return value of the function is multiple names of node .
Please make sure that this is evident from the naming.

An example is as follows:
```suggestion
        node_names_in_mock1, node_names_in_mock2 = Architecture.diff_node_names(architecture_mock1, architecture_mock2)
```

**Review by rokamu623 — CHANGES_REQUESTED** [review]

File: src/test/architecture/test_architecture.py [hunk 7]
```diff
@@ -1019,3 +1019,140 @@ def test_rename_path(self, mocker):
...
+
+    def test_diff_node_topics(self, mocker):
+        architecture_mock1 = mocker.Mock(spec=Architecture)
+        architecture_mock2 = mocker.Mock(spec=Architecture)
+        mocker.patch.object(architecture_mock1, 'topic_names', ('topic1', 'topic2'))
+        mocker.patch.object(architecture_mock2, 'topic_names', ('topic1', 'topic3', 'topic4'))
+        topic_1, topic_2 = Architecture.diff_node_topics(architecture_mock1, architecture_mock2)
+        assert topic_1 == ('topic2',)
+        assert topic_2 == ('topic3', 'topic4') or \
+               topic_2 == ('topic4', 'topic3')
```
Comment (rokamu623) [inline src/test/architecture/test_architecture.py:1060]: In order to adapt to any order, how about the following implementation?

```suggestion
        assert sorted(topic_2) == sorted(('topic3', 'topic4'))
```

**Review by rokamu623 — CHANGES_REQUESTED** [review]

**Comment by taro-yu** [comment]
@rokamu623 
Review points have been corrected. 
Please check.

**Review by rokamu623 — APPROVED** [review]
LGTM!

File: src/caret_analyze/architecture/architecture.py [hunk 8]
```diff
@@ -466,6 +466,41 @@ def rename_topic(self, src: str, dst: str) -> None:
         for c in self._communications:
             c.rename_topic(src, dst)
 
+    @staticmethod
+    def diff_node_names(
```
Comment (rokamu623) [inline src/caret_analyze/architecture/architecture.py:470]: doc string must be described (for each function).

Reference:
https://github.com/tier4/CARET_analyze/blob/f17cd5b74af6d12398e356809368dade5e856675/src/caret_analyze/architecture/architecture.py#L397-L408

**Review by rokamu623 — CHANGES_REQUESTED** [review]

**Comment by taro-yu** [comment]
@rokamu623 
Add doc string for the five diff functions.
please check.

**Comment by taro-yu** [comment]
@rokamu623 
move DiffNode class in architecture,.py to node.py

File: src/caret_analyze/architecture/architecture.py [hunk 9]
```diff
@@ -466,6 +467,101 @@ def rename_topic(self, src: str, dst: str) -> None:
...
+        left_node : NodeStructValue
+            Node in architecture
+        right_node : NodeStructValue
+            Node in architecture
+
+        """
+        return DiffNode(left_node, right_node).diff_node_subs()
+
+    @staticmethod
+    def diff_node_callbacks(
```
[the remark this answers is not in the document]
Comment (taro-yu) [reply src/caret_analyze/architecture/architecture.py]: @takam5f2 
remove the diff_callbacks func and the test codes about it.

File: src/test/architecture/test_architecture.py [hunk 10]
```diff
@@ -1019,3 +1019,143 @@ def test_rename_path(self, mocker):
         assert set(arch.communications) == set(arch_expected.communications)
         assert set(arch.paths) == set(arch_expected.paths)
         assert set(arch.executors) == set(arch_expected.executors)
+
+
+class TestDiff:
```
[the remark this answers is not in the document]
Comment (rokamu623) [reply src/test/architecture/test_architecture.py]: Fixed it in https://github.com/tier4/CARET_analyze/pull/245/commits/c2055136072541d7d00fe85d0dad0d23cf08b5d4 .

File: src/caret_analyze/architecture/architecture.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rokamu623) [reply src/caret_analyze/architecture/architecture.py]: fixed it in https://github.com/tier4/CARET_analyze/pull/245/commits/e595a3cb36c63bd79b3d39d986d22a58ebdd582c .

**Comment by taro-yu** [comment]
@takam5f2 

> @taro-yu Could you add the comment I suggested? This implementation should be temporal in this moment.

added TODO comment in node.py

File: src/caret_analyze/architecture/architecture.py [hunk 11]
```diff
@@ -466,6 +467,82 @@ def rename_topic(self, src: str, dst: str) -> None:
...
             c.rename_topic(src, dst)
 
+    @staticmethod
+    def diff_node_names(
+        left_arch: Architecture,
+        right_arch: Architecture
+    ) -> Tuple[Tuple[str, ...], Tuple[str, ...]]:
+        """
+        Compare two architecture objects (: left_arch, right_arch) and returns\
+        the node names that exist only in the respective architectures.
```
Comment (rokamu623) [inline src/caret_analyze/architecture/architecture.py:477]: For all docstring:

The explanation is redundant.

```suggestion
        Compare two architecture objects and return the difference of nodes name.
```

File: src/caret_analyze/architecture/architecture.py [hunk 12]
```diff
@@ -466,6 +467,82 @@ def rename_topic(self, src: str, dst: str) -> None:
...
+
+        Parameters
+        ----------
+        left_arch : Architecture
+            Architecture object
+        right_arch : Architecture
+            Architecture object
+
```
Comment (rokamu623) [inline src/caret_analyze/architecture/architecture.py:485]: For all docstring:

The "Returns" field is missing.

Reference: 
https://github.com/tier4/CARET_analyze/blob/main/src/caret_analyze/architecture/combine_path.py#L46

**Review by rokamu623 — CHANGES_REQUESTED** [review]
docstring should be revision in 2 point each 3 function.

**Review by rokamu623 — APPROVED** [review]
LGTM!

**Pull request merged**